invokeai/app/invocations/latent.py

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) 2023 Kyle Schouviller (https://github.com/kyle0654)
+
+from typing import Literal, Optional
+from pydantic import BaseModel, Field
+from torch import Tensor
+import torch
+
+from ...backend.stable_diffusion.diffusion.shared_invokeai_diffusion import PostprocessingSettings
+
+from ...backend.image_util.seamless import configure_model_padding
+
+from ...backend.prompting.conditioning import get_uc_and_c_and_ec
+
+from ...backend.stable_diffusion.diffusers_pipeline import ConditioningData, StableDiffusionGeneratorPipeline
+
+from .baseinvocation import BaseInvocation, BaseInvocationOutput, InvocationContext
+
+
+import numpy as np
+
+from PIL import Image
+from skimage.exposure.histogram_matching import match_histograms
+
+from ..services.image_storage import ImageType
+from ..services.invocation_services import InvocationServices
+from .baseinvocation import BaseInvocation, InvocationContext
+from .image import ImageField, ImageOutput
+from ...backend.generator import Txt2Img, Img2Img, Inpaint, InvokeAIGenerator, Generator
+from ...backend.stable_diffusion import PipelineIntermediateState
+from ...backend.util.util import image_to_dataURL
+from diffusers.schedulers import SchedulerMixin as Scheduler
+import diffusers
+from diffusers import DiffusionPipeline
+
+
+class LatentField(BaseModel):
+    """A latent field used for passing latent objects between invocations"""
+
+    latent_name: Optional[str] = Field(default=None, description="The name of the latent")
+
+
+class LatentOutput(BaseInvocationOutput):
+    """Base class for invocations that output an image"""
+    #fmt: off
+    type: Literal["latent"] = "latent"
+    latent:      LatentField = Field(default=None, description="The output latent")
+    #fmt: on
+
+
+# TODO: this seems like a hack
+scheduler_map = dict(
+    ddim=diffusers.DDIMScheduler,
+    dpmpp_2=diffusers.DPMSolverMultistepScheduler,
+    k_dpm_2=diffusers.KDPM2DiscreteScheduler,
+    k_dpm_2_a=diffusers.KDPM2AncestralDiscreteScheduler,
+    k_dpmpp_2=diffusers.DPMSolverMultistepScheduler,
+    k_euler=diffusers.EulerDiscreteScheduler,
+    k_euler_a=diffusers.EulerAncestralDiscreteScheduler,
+    k_heun=diffusers.HeunDiscreteScheduler,
+    k_lms=diffusers.LMSDiscreteScheduler,
+    plms=diffusers.PNDMScheduler,
+)
+
+
+SAMPLER_NAME_VALUES = Literal[
+    tuple(list(scheduler_map.keys()))
+]
+
+
+def get_scheduler(scheduler_name:str, model: StableDiffusionGeneratorPipeline)->Scheduler:
+    scheduler_class = scheduler_map.get(scheduler_name,'ddim')
+    scheduler = scheduler_class.from_config(model.scheduler.config)
+    # hack copied over from generate.py
+    if not hasattr(scheduler, 'uses_inpainting_model'):
+        scheduler.uses_inpainting_model = lambda: False
+    return scheduler
+
+
+# TODO: is this a system setting?
+def torch_dtype(precision:str = "float16") -> torch.dtype:
```

Comment:
i think you want to be using `torch_dtype()` in `ldm.invoke.devices.py`

Reply:
Moved to `invokeai.backend.util.devices` but otherwise unchanged.